### PR TITLE
Python AIOHTTP server generator

### PR DESCRIPTION
**Pull request by Jyhess**
@wing328 @taxpon @frol @mbohlool @cbornet @kenjones-cisco @tomplus

Following issue [1322](https://github.com/OpenAPITools/openapi-generator/issues/1322), I started to write a generator for aiohttp server.
It's a work in progress, and I would like early review and help to fix some remaining issues.

**Note**:
- I think some improvements can be applied to flask generator (like security management for example), but for the moment, I didn't change it as I'm not familiar with.

And here is some questions:

In [preprocessOpenAPI function](https://github.com/Jyhess/openapi-generator/blob/aiohttp_server_generator/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonAbstractConnexionServerCodegen.java#L389), I set body name to "body" by default. I would like set it to type name (_pet_ or _user_ is better, but didn't figure how to get easily this information. Do you know how I can find this information ?

In [model.mustach](https://github.com/Jyhess/openapi-generator/blob/aiohttp_server_generator/modules/openapi-generator/src/main/resources/aiohttpConnexion/model.mustache), I would like to generate imports for other generated class, like
```
from openapi_server.models.category import Category
from openapi_server.models.tag import Tag
```
Do you know how I can do that ? There is the same issue with Flask generator.

In [test controllers](https://github.com/Jyhess/openapi-generator/blob/aiohttp_server_generator/modules/openapi-generator/src/main/resources/aiohttpConnexion/controller_test.mustache#L18), examples are not generated. Any clue on that ?

Do you know why side files, like [git_push.sh](https://github.com/Jyhess/openapi-generator/blob/aiohttp_server_generator/samples/server/petstore/aiohttpConnexion/git_push.sh), or [setup.py](https://github.com/Jyhess/openapi-generator/blob/aiohttp_server_generator/samples/server/petstore/aiohttpConnexion/setup.py) are generated? I'm not sure they should be part of a code generator.

### CODE REVIEW

**Comment by wing328** [comment]
Thanks for the PR but your commit (as shown in the Commits tab) is not linked to your Github account, which means this PR won't count as your contribution in https://github.com/OpenAPITools/openapi-generator/graphs/contributors.

Let me know if you need help fixing it.

Ref: https://github.com/OpenAPITools/openapi-generator/wiki/FAQ#how-can-i-update-commits-that-are-not-linked-to-my-github-account

**Comment by wing328** [comment]
@Jyhess thanks for the PR. Are you free on the coming week to have a quick chat about this PR?

I'm available via https://gitter.im (ID: wing328) or Google Hangout (you can find the email address in my profile page)

**Comment by Jyhess** [comment]
Hi William,

We can have a chat on gitter this week.
I'm available from 10 am french timezone, 5 pm for you I guess.

FYI, I updated my PR with some more fixes.

BR,
Julien


Le sam. 24 nov. 2018 à 10:18, William Cheng <notifications@github.com> a
écrit :

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonAbstractConnexionServerCodegen.java [hunk 1]
```diff
@@ -0,0 +1,845 @@
+/*
+ * Copyright 2018 OpenAPI-Generator Contributors (https://openapi-generator.tech)
+ * Copyright 2018 SmartBear Software
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/PythonAbstractConnexionServerCodegen.java]: @Jyhess for new file added to this project, we'll no longer add the 3rd line. Please refer to https://opensource.stackexchange.com/questions/7300/copyright-notice-in-the-file-header-apache-v2-license/7301#7301 for more information.

File: bin/python-aiohttp-petstore.sh [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+executable="./modules/openapi-generator-cli/target/openapi-generator-cli.jar"
+
+if [ ! -f "$executable" ]
+then
+  mvn -B clean package
+fi
+
+# if you've executed sbt assembly previously it will use that instead.
+export JAVA_OPTS="${JAVA_OPTS} -XX:MaxPermSize=256M -Xmx1024M -DloggerPath=conf/log4j.properties"
+ags="generate -t modules/openapi-generator/src/main/resources/aiohttpConnexion -i modules/openapi-generator/src/test/resources/2_0/petstore.yaml -g python-aiohttp -o samples/server/petstore/aiohttpConnexion -Dservice $@"
```
Comment (wing328) [inline bin/python-aiohttp-petstore.sh]: For the resource folder, it's usually the same name as the generator `python-aiohttp`.

**Comment by wing328** [comment]
@Jyhess currently, we test the python clients in Travis by adding the folders to pom.xml: https://github.com/OpenAPITools/openapi-generator/blob/master/pom.xml#L1039

If it's not a lot of work, I suggest we do the same for Python AIOHTTP Petstore server samples by adding a [pom.xml](https://github.com/OpenAPITools/openapi-generator/blob/master/samples/client/petstore/python/pom.xml) under `samples/server/petstore/aiohttpConnexion/`

File: bin/python-aiohttp-petstore.sh (hunk 2 above)
Comment (Jyhess) [reply bin/python-aiohttp-petstore.sh]: Done

**Comment by Jyhess** [comment]
Tests for generated server cannot succeed until [this PR](https://github.com/zalando/connexion/pull/760) on connexion is not released.

Moreover, there is some other issues in generated tests for the moment:
- examples are empty, which occur validation error for mandatory fields
- form and file upload are not yet managed
- body with list are not yet supported
I will try to fix them, but I don't know when.

I will add travis files and find a way to deactivate or fix failing tests.

**Comment by Jyhess** [comment]
@wing328 I fixed tests. I also added and fixed tests for python-flask generator.

**Comment by wing328** [comment]
I'll review again later today or tomorrow and merge it into master if no further feedback

**Comment by wing328** [comment]
Tweet: https://twitter.com/oas_generator/status/1083751186782552064

HN: https://news.ycombinator.com/item?id=18884281

Reddit: https://www.reddit.com/r/programming/comments/aewn74/generate_a_python_aiohttp_server_with_openapi/

**Pull request merged**